### ISSUE TITLE
Placeholder with shape=(None, 1024) gives an error when used with tf.cond

### ISSUE THREAD

**Issue by zeis** [tensorflow/tensorflow#11371]
I'm using TensorFlow v1.2.0. Here is a simple example that shows the problem.
If a dimension of a placeholder is None I get the following error:

```
InvalidArgumentError (see above for traceback): Shape [-1,1024] has negative dimensions
	 [[Node: cond/Placeholder = Placeholder[dtype=DT_FLOAT, shape=[?,1024], _device="/job:localhost/replica:0/task:0/cpu:0"](^cond/switch_t)]]
```

If I set the shape to None or to (1024, 1024) it gives no error. 

```
import numpy as np
import tensorflow as tf

rand_array = np.random.rand(1024, 1024)

x_c = tf.constant(rand_array, dtype=tf.float32)

py_flag = False
flag = tf.constant(py_flag, tf.bool)

def pl():
    return tf.placeholder(dtype=tf.float32, shape=(None, 1024))

x = tf.cond(flag, pl, lambda: x_c)

y = tf.matmul(x, x)

with tf.Session() as sess:
    if py_flag:
        print(sess.run(y, feed_dict={x: rand_array}))
    else:
        print(sess.run(y))
```

**Comment by jowettcz** [comment]
if py_flag == True, then session run. But palceholder cannot find the feed_dict. Then error happened.
There is possibility in other places you call the op without feed_dict. Have a try, good look.

**Comment by michaelisard** [comment]
It looks as if this is fixed in the master branch: would you try there?

**Comment by zeis** [comment]
I'm having troubles installing the latest version of Bazel unfortunately. It only let me install version 0.1.3 and I can't build TensorFlow as it requires version 0.4.5 (on a Mac). I would suggest though that if you tested my example and it works in the master branch then we can mark this as closed. Thank you.

**Comment by oborchers** [comment]
I am having a very similar problem to this one. Tf cond does not seem to affect my input placeholder to which I want to ad some salt and pepper noise. Consider the following code:

`
def noise_mask(input, keep_prob):    
    return tf.zeros_like(input)

def binary_noise(input, keep_prob=0.05, is_training=True):
    is_training = tf.Variable(is_training)
    return tf.cond(is_training, lambda: noise_mask(input, keep_prob), lambda: tf.identity(input))
`
This simple version sets all tensor values to zero if desired, i.e. if is training == true.
Then I apply this to the output layer:
`self.out = utils.binary_noise(tf.add(tf.matmul(self.hidden, self.weights['w2']), self.weights['b2']), is_training=True)`

As expected, the loss stagnates as the output layer is not able to pass the gradient through.
`Epoch:   1      Cost:0.693194389        Loss:0.692861199
Epoch:   2      Cost:0.693194389        Loss:0.692861199
Epoch:   3      Cost:0.693194389        Loss:0.692861199`

Now, I would like to apply this function to the input to add binary noise.
Consider:
`
self.x_input = tf.placeholder(tf.float64, [None, self.n_input])
self.x = utils.binary_noise(self.x_input, keep_prob=0.05, is_training=False)`

However the loss changes as well, when is_training == True. I have no idea why this is the case.

**Comment by oborchers** [comment]
I am able to replicate the error of the first post.
    self.x_input = tf.placeholder(tf.float64, [None, self.n_input])
    self.x = utils.binary_noise(self.x_input, self.is_training)

As I change "None" for 100 everything works fine.

InvalidArgumentError (see above for traceback): Shape [-1,2000] has negative dimensions
	 [[Node: AEC/Placeholder_1 = Placeholder[dtype=DT_DOUBLE, shape=[?,2000], _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

**Comment by jart** [comment]
@zeis If you have trouble getting Bazel on your system, you could always try installing the nightly. For example, one such URL: `pip install --upgrade https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON2,label=cpu-slave/lastSuccessfulBuild/artifact/pip_test/whl/tensorflow-1.2.1-cp27-none-linux_x86_64.whl`. Could you please confirm that it solves the problem for you?

**Comment by zeis** [comment]
Thanks @jart.

I've just tested it using the following nightly build and it worked!

https://ci.tensorflow.org/view/Nightly/job/nightly-matrix-cpu/548/TF_BUILD_IS_OPT=OPT,TF_BUILD_IS_PIP=PIP,TF_BUILD_PYTHON_VERSION=PYTHON3,label=mac-slave/artifact/pip_test/whl/tensorflow-1.2.1-cp35-cp35m-macosx_10_11_x86_64.whl

**Comment by jart** [comment]
That's great to hear; thank you for confirming. I'll close this one out now.

**Comment by Mistobaan** [comment]
@jart I would add a test case to the repository to confirm that the issue is solved and to have it in the future for regression testing, before closing the issue itself.
